pkg/auth/service.go

```diff
@@ -471,478 +518,494 @@ func (s *DBAuthService) ListEffectivePolicies(ctx context.Context, username stri
-	policy, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return getPolicy(tx, policyDisplayName)
-	}, db.ReadOnly())
+func (s *KVAuthService) GetPolicy(ctx context.Context, policyDisplayName string) (*model.BasePolicy, error) {
+	policyKey := model.PolicyPath(policyDisplayName)
+	p := model.PolicyData{}
+	_, err := s.store.GetMsg(ctx, model.PartitionKey, policyKey, &p)
 	if err != nil {
-		return nil, err
+		if errors.Is(err, kv.ErrNotFound) {
+			err = ErrNotFound
+		}
+		return nil, fmt.Errorf("%s: %w", policyDisplayName, err)
 	}
-	return &policy.(*model.DBPolicy).BasePolicy, nil
+	return model.PolicyFromProto(&p), nil
 }
 
-func (s *DBAuthService) DeletePolicy(ctx context.Context, policyDisplayName string) error {
-	_, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return nil, deleteOrNotFound(tx, `DELETE FROM auth_policies WHERE display_name = $1`, policyDisplayName)
-	})
-	return err
-}
+func (s *KVAuthService) DeletePolicy(ctx context.Context, policyDisplayName string) error {
+	if _, err := s.GetPolicy(ctx, policyDisplayName); err != nil {
+		return err
+	}
+	policyPath := model.PolicyPath(policyDisplayName)
 
-func (s *DBAuthService) ListPolicies(ctx context.Context, params *model.PaginationParams) ([]*model.BasePolicy, *model.Paginator, error) {
-	type res struct {
-		policies  []*model.BasePolicy
-		paginator *model.Paginator
-	}
-	result, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		dbPolicies := make([]*model.DBPolicy, 0)
-		err := tx.Select(&dbPolicies, `
-			SELECT *
-			FROM auth_policies
-			WHERE display_name > $1
-			AND display_name LIKE $2
-			ORDER BY display_name
-			LIMIT $3`,
-			params.After, fmt.Sprint(params.Prefix, "%"), params.Amount+1)
-		if err != nil {
-			return nil, err
-		}
-		p := &model.Paginator{}
-		policies := make([]*model.BasePolicy, len(dbPolicies))
-		for i := range dbPolicies {
-			policies[i] = &dbPolicies[i].BasePolicy
+	// delete policy attachment to user
+	usersKey := model.UserPath("")
+	it, err := kv.NewPrimaryIterator(ctx, s.store.Store, (&model.UserData{}).ProtoReflect().Type(), model.PartitionKey, usersKey, "")
+	if err != nil {
+		return err
+	}
+	defer it.Close()
+	for it.Next() {
+		entry := it.Entry()
+		user := entry.Value.(*model.UserData)
+		if err = s.DetachPolicyFromUserNoValidation(ctx, policyDisplayName, user.Username); err != nil {
+			return err
 		}
-		if len(policies) == params.Amount+1 {
-			// we have more pages
-			policies = policies[0:params.Amount]
-			p.Amount = params.Amount
-			p.NextPageToken = policies[len(policies)-1].DisplayName
-			return &res{policies, p}, nil
+	}
+
+	// delete policy attachment to group
+	groupKey := model.GroupPath("")
+	it, err = kv.NewPrimaryIterator(ctx, s.store.Store, (&model.UserData{}).ProtoReflect().Type(), model.PartitionKey, groupKey, "")
+	if err != nil {
+		return err
+	}
+	defer it.Close()
+	for it.Next() {
+		entry := it.Entry()
+		group := entry.Value.(*model.GroupData)
```

Comment:
it initialized with type `UserData` and attempted to be accessed as `GroupData`. This wiil crash

Reply:
done